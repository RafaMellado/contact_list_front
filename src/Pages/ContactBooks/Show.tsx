import { useEffect, useState } from "react";
import { Button, Col, Row } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import { useNavigate, useParams } from "react-router-dom";
import ContactBooksService from "../../Services/ContactBooksService";
import { ContactBookWithContacts } from "../../Services/Interfaces/ContactBook";
import { ContactBookContactCard } from "../../Components/ContactBookContactCard";
import ContactsService from "../../Services/ContactsService";

export function ContactBookShow() {
  const [contactBook, setContactBook] = useState<ContactBookWithContacts>();

  const { id } = useParams();
  const { t } = useTranslation();

  const navigate = useNavigate();

  const TRANSLATIONS: string = "contactBooksShow";

  useEffect(() => {
    ContactBooksService.show(Number(id)).then((response) =>
      setContactBook(response)
    );
  }, [id]);

  const addContact = () => {
    navigate(`/contact-book/${id}/contact/new`);
  };

  const backBtn = () => {
    navigate("/home");
  };

  const goToContactShow = (contactId: number) => {
    navigate(`/contact/${contactId}`);
  };

  const goToEditContact = (contactId: number) => {
    navigate(`/contact/${contactId}/edit`);
  };

  const deleteContact = async (id: number) => {
    await ContactsService.delete(id);

    reassignContacts(id);
  };

  const reassignContacts = (deletedId: number) => {
    const updatedContactBook = {
      ...contactBook,
      contacts: contactBook?.contacts.filter((item) => item.id !== deletedId),
    };

    setContactBook(updatedContactBook as ContactBookWithContacts);
  };

  return (
    <Row>
      <Col>
        <h2>{contactBook?.name}</h2>

        <Button className="me-2" onClick={backBtn}>
          {t<string>(`${TRANSLATIONS}.backBtn`)}
        </Button>

        <Button onClick={addContact}>
          {t<string>(`${TRANSLATIONS}.addContactBtn`)}
        </Button>

        <h3>{t<string>(`${TRANSLATIONS}.contacts`)}</h3>

        <Row>
          {contactBook?.contacts.map((item) => {
            return (
              <Col xs={12} sm={6} md={3} key={item.id}>
                <ContactBookContactCard
                  item={item}
                  cardFn={() => goToContactShow(item.id)}
                  editFn={() => goToEditContact(item.id)}
                  deleteFn={() => deleteContact(item.id)}
                />
              </Col>
            );
          })}
        </Row>
      </Col>
    </Row>
  );
}
