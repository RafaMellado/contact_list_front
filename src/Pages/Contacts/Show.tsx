import { useEffect, useState } from "react";
import { Button, Col, Row } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import { useNavigate, useParams } from "react-router-dom";
import ContactsService from "../../Services/ContactsService";
import { ContactWithContactHistories } from "../../Services/Interfaces/Contact";
import { ContactCard } from "../../Components/ContactCard";
import { ContactHistoryCard } from "../../Components/ContactHistoryCard";

export function ContactShow() {
  const [contact, setContact] = useState<ContactWithContactHistories>();

  const { id } = useParams();
  const { t } = useTranslation();

  const navigate = useNavigate();

  const TRANSLATIONS: string = "contactsShow";

  useEffect(() => {
    ContactsService.show(Number(id)).then((response) => setContact(response));
  }, [id]);

  const backBtn = (id: number) => {
    navigate(`/contact-book/show/${id}`);
  };

  const editContact = () => {
    navigate(`/contact/${id}/edit`);
  };

  return (
    <Row>
      <Col xs={12} md={6}>
        {contact && (
          <ContactCard item={contact}>
            <>
              <Button
                className="me-2"
                onClick={() => backBtn(Number(contact.contact_book_id))}
              >
                {t<string>(`${TRANSLATIONS}.backBtn`)}
              </Button>

              <Button className="me-2" onClick={editContact}>
                {t<string>(`${TRANSLATIONS}.editContact`)}
              </Button>
            </>
          </ContactCard>
        )}
      </Col>

      <Col xs={12} md={6}>
        {!!contact?.contact_histories.length && (
          <>
            <h2>{t<string>(`${TRANSLATIONS}.history`)}</h2>

            {contact.contact_histories.map((item, index) => {
              return (
                <ContactHistoryCard
                  item={item}
                  key={index}
                  translations={TRANSLATIONS}
                />
              );
            })}
          </>
        )}
      </Col>
    </Row>
  );
}
