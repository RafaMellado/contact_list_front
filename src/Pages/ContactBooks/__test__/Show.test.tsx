import { ContactBookShow } from "../Show";
import { act, render } from "@testing-library/react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

describe(`<${ContactBookShow.name} />`, () => {
  const history = createMemoryHistory();

  const factoryComponent = () =>
    render(
      <Router location="/" navigator={history}>
        <ContactBookShow />
      </Router>
    );

  test("renders correctly the component", () => {
    const component = factoryComponent();

    act(() => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });
});
