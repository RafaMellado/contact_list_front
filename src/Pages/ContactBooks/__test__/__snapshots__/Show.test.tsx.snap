// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactBookShow /> renders correctly the component 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <button
        class="me-2 btn btn-primary"
        type="button"
      >
        contactBooksShow.backBtn
      </button>
      <div
        class="d-flex align-items-center justify-content-between my-4 flex-column flex-md-row"
      >
        <h2>
          contactBooksShow.title
        </h2>
        <button
          class="mt-3 mt-md-0 btn btn-primary"
          type="button"
        >
          contactBooksShow.addContactBtn
        </button>
      </div>
      <div
        class="mb-4"
      >
        <input
          placeholder="contactBooksShow.search"
        />
      </div>
      <div
        class="row"
      />
    </div>
  </div>
</DocumentFragment>
`;
